### PR TITLE
feat: [PIE-5849]: NG - Implement authorization for custom webhooks

### PR DESCRIPTION
**Pull request by vcalasansh**
## Describe your changes
This PR implements authorization for custom webhooks in NG.

By default, if the user passes `X-Api-Key` header in the call to `/webhook/custom` or `/webhook/custom/v2` the API Key will always be validated.

if `NG_SETTINGS` Feature Flag is enabled, and the account/org/project setting  `mandate_custom_webhook_authorization` is set to `true`, then authorization (and `X-Api-Key` in header) will be mandatory.

**Testing instructions**
 - Head to the PR env for this feature: https://spg.pr.harness.io/pie-5849-11/.
 - Login with admin account. Create a pipeline and custom trigger for it.
 - Create an API key for the user in "profile" section.
 - Call the webhook with `curl` passing `X-Api-Key` header. This will answer with "SUCCESS".
 - Log in to the account with e-mail: readonlyuser@harness.io, password: readonlyuser.
 - Create an API key for this user.
 - Try to call the webhook again with readonlyuser's API Key in the `X-API-Key` header. This will fail with code `NG_ACCESS_DENIED`.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39011)

### CODE REVIEW

**Review by vaibhav-si — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39011)* status: 13 file(s) to review. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/annotations/PipelineServiceAuthIfHasApiKey.java` line 24 at r1](https://reviewable.io/reviews/harness/harness-core/39011#-NGVL2-6B3kSLrZTVnYS:-NGVL2-6B3kSLrZTVnYT:b-isohkw) ([raw file](https://github.com/harness/harness-core/blob/baa44a23a364e6e557a069c572abc2248e13f348/pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/annotations/PipelineServiceAuthIfHasApiKey.java#L24)):*
> ```Java
> @Retention(RetentionPolicy.RUNTIME)
> @Target({ElementType.METHOD, ElementType.TYPE})
> public @interface PipelineServiceAuthIfHasApiKey {}
> ```

pls add some description to this

**Review by vcalasansh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39011)* status: 13 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/annotations/PipelineServiceAuthIfHasApiKey.java` line 24 at r1](https://reviewable.io/reviews/harness/harness-core/39011#-NGVL2-6B3kSLrZTVnYS:-NGX0wYBCf4tJO9Py6Lw:bf3maur) ([raw file](https://github.com/harness/harness-core/blob/baa44a23a364e6e557a069c572abc2248e13f348/pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/annotations/PipelineServiceAuthIfHasApiKey.java#L24)):*
<details><summary><i>Previously, vaibhav-si (Vaibhav Singhal) wrote…</i></summary><blockquote>

pls add some description to this
</blockquote></details>

Done, thank you!

File: pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/annotations/PipelineServiceAuthIfHasApiKey.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+// This annotation can be used in a resource to validate the X-Api-Key header if (and only if) it is sent
+// in the request headers.
+@OwnedBy(SPG)
+@Inherited
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD, ElementType.TYPE})
```
Comment (sribalij) [inline pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/annotations/PipelineServiceAuthIfHasApiKey.java:25]: Do we need it at the class level or just ElementType.METHOD is enough?

**Review by vcalasansh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39011)* status: 14 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java` line 1010 at r2](https://reviewable.io/reviews/harness/harness-core/39011#-NEm3FRm3g4sgXukqA62-r2-1010:-NGYHy3a3OK_cnexoZXX:b-t3l1bm) ([raw file](https://github.com/harness/harness-core/blob/df1ebf0df3592411dc75fac412e44910d11d4a66/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java#L1010)):*
<details><summary><i>Previously, sribalij (Sridhar) wrote…</i></summary><blockquote>

Add a Static String
</blockquote></details>

Done, thank you!!
___
*[`pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/annotations/PipelineServiceAuthIfHasApiKey.java` line 25 at r2](https://reviewable.io/reviews/harness/harness-core/39011#-NEm3FRm3g4sgXukqA63-r2-25:-NGYHzkTDJZqyeSuhqLb:b-2hcw1j) ([raw file](https://github.com/harness/harness-core/blob/df1ebf0df3592411dc75fac412e44910d11d4a66/pipeline-service/modules/pms-contracts/src/main/java/io/harness/pms/annotations/PipelineServiceAuthIfHasApiKey.java#L25)):*
<details><summary><i>Previously, sribalij (Sridhar) wrote…</i></summary><blockquote>

Do we need it at the class level or just ElementType.METHOD is enough?
</blockquote></details>

ElementType.METHOD is enough. I updated it, thank you!

**Review by sribalij — APPROVED** [review]
> Reviewed 10 of 13 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39011)* status: 14 file(s) to review. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> Reviewed 10 of 13 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39011)* status: 3. unresolved discussions.

**Review by sribalij — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39011)* status: 1. unresolved discussions.

**Comment by vcalasansh** [comment]
trigger ti0

**Pull request closed without merging**